lib/ssl/src/ssl_handshake.erl

```diff
@@ -2148,6 +2148,12 @@ digitally_signed(Version, Msg, HashAlgo, PrivateKey, SignAlgo) ->
 	    throw(?ALERT_REC(?FATAL, ?HANDSHAKE_FAILURE, bad_key(PrivateKey)))
     end.
 
+do_digitally_signed(Version, Msg, HashAlgo, #{algorithm := _KeyAlg, 
+                                              sign_fun := Sign} = Key, SignAlgo) when is_function(Sign) ->
+    Sign(Version, Msg, HashAlgo, Key, SignAlgo);
+do_digitally_signed(Version, Msg, HashAlgo, #{algorithm := _KeyAlg,
+                                              sign_fun := {Mod, Fun}} = Key, SignAlgo) ->
+    Mod:Fun(Version, Msg, HashAlgo, Key, SignAlgo);
```

Comment:
@IngelaAndin Ok I can try moving the handling of `sign_fun` in `public_key`.
But your second options sound like what is already here, or are you saying to move even the calls to crypto into `public_key`?

Reply:
So public key is the more general API. That is why I think it
would be better if it is the place that handles different type of
keys. Although engine keys are very OpenSSL specific and hence
will only be a passthrough to crypto. When Engine was a special
case that did not fit under the general case that was that all
general keys where supplied as DER (raw) or through PEM-format
the reason to go through public_key was not obvious. But now
engine becomes a corner case of the more general case of suppling
a fun. And this may be something that is interesting to other
users, than ssl, of the public_key API.

To support old TLS-versions (pre TLS-1.2) we will also need
support for the encrypt/decrypt_private that only will need to
handle plain RSA keys (with param = NULL).